### ISSUE TITLE
Invoke-Command with ErrorAction SilentlyContinue still throws Exception

### ISSUE THREAD

**Issue by deathly809** [PowerShell/PowerShell#7623]
When running a script using Invoke-Command I call a program which prints to the error stream.  I have piped the error stream to the stdout and then Out-Null for the command I call and also set the ErrorAction to SilentlyContinue for Invoke-Command.  The exception is still propagated.  

Steps to reproduce
------------------

Using the following files,

[error.ps1.txt](https://github.com/PowerShell/PowerShell/files/2315989/error.ps1.txt)
[cradle.ps1.txt](https://github.com/PowerShell/PowerShell/files/2315990/cradle.ps1.txt)

```powershell
.\cradle.ps1
```

Expected behavior
-----------------

```none
Running
Before
After
Finished
```

Actual behavior
---------------

```none
Running

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    CmdletBindingAttribute                   System.Management.Automation.CmdletCommonMetadataAttribute
Before
There was an error




PSMessageDetails      :
Exception             : System.Management.Automation.RemoteException: '1' is not recognized as an internal or external
                        command,
TargetObject          : '1' is not recognized as an internal or external command,
CategoryInfo          : NotSpecified: ('1' is not reco...ternal command,:String) [], RemoteException
FullyQualifiedErrorId : NativeCommandError
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, C:\Users\Administrator.N22R1003-DVM\Desktop\Terasort\Error\error.ps1: line 7
                        at DoIt, C:\Users\Administrator.N22R1003-DVM\Desktop\Terasort\Error\error.ps1: line 6
                        at <ScriptBlock>, C:\Users\Administrator.N22R1003-DVM\Desktop\Terasort\Error\cradle.ps1: line
                        10
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}



Finished
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      5.1.14393.2395
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.2395
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

**Comment by MaximoTrinidad** [comment]
First, you need to try it on PowerShell Core, as this is only for PowerShell Core issues.  So, for Windows PowerShell issues, use Uservoice at: https://windowsserver.uservoice.com/forums/301869-powershell 

Overall, this type of exception can only be handle using the try/catch.  

In order to reproduce in PowerShell Core, we need better steps. We don't know what commands are executing in "Cradle.ps1" script file.

Keep in mind, Microsoft efforts are to move forward with PowerShell Core.

:)

**Comment by deathly809** [comment]
You cannot tell what is in a text file?  I am very confused.  Have you tried notepad?  

To help I have opened the files in notepad and copied the text files here.

```
#cradle.ps1

# Set the error action to Stop
$ErrorActionPreferenceOld = $ErrorActionPreference
$ErrorActionPreference = "Stop"

. .\error.ps1

try {
     Write-Host "Running"
     DoIt
} catch {
    Write-Host "There was an error"
    Format-list * -InputObject $_ -Force
} finally {
     Write-Host "Finished"
     $ErrorActionPreference = $ErrorActionPreferenceOld
}
```

```
#error.ps1

function DoIt {
    [CmdletBinding]
    Write-Host "Before"
    # Perform action which prints to error stream
    Invoke-Command -Script {
        & cmd.exe /k "echo 'test' 1&>2" *>1 | Out-Null
    } -ErrorAction SilentlyContinue | Out-Null -ErrorAction SilentlyContinue
    Write-Host "After"
}
```

Just in case you are confused you will need to copy AND paste these into the following files and please read the first line of the code as those correspond to the files that you need to copy and paste into,

```
cradle.ps1
error.ps1
```

then at the command line please type,

```
.\cradle.ps1
```

Please let me know if that was not clear.  


I have also tried in a try catch and it still throws in my real script and not this toy example and it still fails.

I reproduced it in PowerShell Core, here is the version I have.  I will also file a bug with the PowerShell team.

```
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.4
PSEdition                      Core
GitCommitId                    6.1.0-preview.4
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by BrucePay** [comment]
Here's a reduced version of the code with just the relevant bits.
```powershell
$ErrorActionPreference="silentlycontinue"
Invoke-Command -Script {
        & cmd.exe /c 1 *>1 | Out-Null
} -ErrorAction SilentlyContinue
```
This will indeed generate a error despite setting `-ErrorAction` on `Invoke-Command`. This is because `$ErrorActionPreference` is set to `Stop`. The preference variable is the ambient setting for all commands vs. the `-ErrorAction` parameter that controls the behaviour for a specific command. This means that when cmd.exe is executed, it's error action will be `Stop` causing an exception to be thrown. As @MaximoTrinidad points out, you could capture that exception with `try/catch`. On the other hand, you could change the setting of `$ErrorActionPreference` to be something other than `Stop`. You could even set it inside the `Invoke-Command` scriptblock so it's local to that block:
```powershell
$ErrorActionPreference="stop"
Invoke-Command -Script {
	$ErrorActionPreference="silentlycontinue"
        & cmd.exe /c 1 *>1 | Out-Null
} -ErrorAction SilentlyContinue
```
Ultimately there is no PowerShell bug here - everything is working as intended - however I'd encourage you to look at your overall error handling approach since you appear to have conflicting preferences.

**Comment by MaximoTrinidad** [comment]
Ah! Apologoies @deathly809 !!  

I didn't notice you included other files.  Hehe!!

Thanks BruceP for assisting!!

**Comment by deathly809** [comment]
Hmmm, I will override the setting and see what happens.  Hopefully that fixes it.  Thanks!

**Comment by fatherofinvention** [comment]
Hey guys, this does not seem to be fixed. Is this by design? Invoke-Command does not respect -ErrorAction parameter which means it cannot be suppressed using SilentlyContinue nor can it be trapped using aTry..Catch block.

This issue is present in both WinPS and PowerShell Core. Does not seem like intended behavior since the $Error[0] variable and $ResolutionError ErrorVariable are set by PowerShell outside of the scriptblock.

`Invoke-Command -ScriptBlock { Get-Content c:\file\does\not\exist.txt } -ErrorAction SilentlyContinue -ErrorVariable ResolutionError`

**Comment by hallipr** [comment]
For anyone else still seeing this, ErrorAction on Invoke-Command doesn't affect the error action inside the block you pass to Invoke-Command, just the error action for Invoke-Command and its logic.

If you're passing a ScriptBlock into a function in another context, $ErrorActionPrefence inside the script block will not be affected by $ErrorActionPreference in the function that invokes the script block.  $ErrorActionPreference appears to behave like any other scoped variable, so changing it in the calling function doesn't affect the value captured by the script block.

### With ExecuteBlock defined in the script 
``` powershell
function ExecuteBlock([ScriptBlock]$ScriptBlock) {
    Write-Host "  ExecuteBlock start: $ErrorActionPreference"
    
    $ErrorActionPreference = "Continue"
    Write-Host "  ExecuteBlock set: $ErrorActionPreference"
    
    Invoke-Command -ScriptBlock $ScriptBlock
    Write-Host "  ExecuteBlock end: $ErrorActionPreference"
}

Write-Host "Script start: $ErrorActionPreference"

ExecuteBlock { Write-Host "    ScriptBlock: $ErrorActionPreference" }

Write-Host "Script end: $ErrorActionPreference"
```

```
Script start: Stop
  ExecuteBlock start: Stop
  ExecuteBlock set: Continue
    ScriptBlock: Continue
  ExecuteBlock end: Continue
Script end: Stop
```

### With ExecuteBlock defined in a module.
``` powershell
function ExecuteBlock([ScriptBlock]$ScriptBlock) {
    Write-Host "  ExecuteBlock start: $ErrorActionPreference"
    
    $ErrorActionPreference = "Continue"
    Write-Host "  ExecuteBlock set: $ErrorActionPreference"
    
    Invoke-Command -ScriptBlock $ScriptBlock
    Write-Host "  ExecuteBlock end: $ErrorActionPreference"
}

Export-ModuleMember -Alias * -Function *
```

``` powershell
Import-Module .\helpers.psm1 -Force

Write-Host "Script start: $ErrorActionPreference"

ExecuteBlock { Write-Host "    ScriptBlock: $ErrorActionPreference" }

Write-Host "Script end: $ErrorActionPreference"
```

```
Script start: Stop
  ExecuteBlock start: Stop
  ExecuteBlock set: Continue
    ScriptBlock: Stop
  ExecuteBlock end: Continue
Script end: Stop
```

The module example proves that setting $ErrorActionPreference before calling Invoke-Command will only affect the ErrorActionPreference in the script block if the script block and the invocation are in the same scope.

### Using InvokeWithContext
To invoke a script block from another scope while altering its ErrorActionPreference, you can call InvokeWithContext on the ScriptBlock

https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.scriptblock.invokewithcontext?view=powershellsdk-7.0.0

``` powershell
function ExecuteBlock([ScriptBlock]$ScriptBlock) {
    Write-Host "  ExecuteBlock start: $ErrorActionPreference"

    $ScriptBlock.InvokeWithContext($null, @([PSVariable]::new('ErrorActionPreference', 'Continue')))
    
    Write-Host "  ExecuteBlock end: $ErrorActionPreference"
}
```

```
Script start: Stop
  ExecuteBlock start: Stop
    ScriptBlock: Continue
  ExecuteBlock end: Stop
Script end: Stop
```

## Summing Up

``` powershell
function ExecuteBlock([ScriptBlock]$ScriptBlock) {
    $ScriptBlock.InvokeWithContext($null, @([PSVariable]::new('ErrorActionPreference', 'Continue')))
    Write-Host "ExecuteBlock `$? = $?"
    Write-Host "ExecuteBlock `$LASTEXITCODE = $LASTEXITCODE"
}

Export-ModuleMember -Alias * -Function *
```

``` powershell
Import-Module .\helpers.psm1 -Force

ExecuteBlock { & cmd.exe /c "1" *>&1 }

Write-Host "Script `$?: $?"
```

```
'1' is not recognized as an internal or external command,
operable program or batch file.
ExecuteBlock $? = True
ExecuteBlock $LASTEXITCODE = 1
Script $?: True
```
